File: android/src/main/kotlin/br/com/dextra/bifrost/BifrostNotificationChannel.kt
package br.com.dextra.bifrost

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal object BifrostNotificationChannel : MethodChannel.MethodCallHandler {

  private val notifications = hashMapOf<String, BifrostNotificationCallback>()
  private val channel = MethodChannel(Bifrost.engine.dartExecutor, "bifrost/notification")

  init {
    channel.setMethodCallHandler(this)
  }

  fun register(key: String, callback: (arguments: Any?) -> Unit) {
    notifications[key] = object : BifrostNotificationCallback {
      override fun onReceiveNotification(arguments: Any?) {
        callback.invoke(arguments)
      }
    }
  }

  fun register(key: String, callback: BifrostNotificationCallback) {
    notifications[key] = callback
  }

  fun unregister(key: String) {
    notifications.remove(key)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    val key = call.method
    val args = call.arguments
    notifications[key]?.onReceiveNotification(args)

    result.success(notifications.containsKey(key))
  }
}

interface BifrostNotificationCallback {

  fun onReceiveNotification(arguments: Any?)
}

File: ios/Classes/BifrostNotificationChannel.swift
//
//  BifrostNotificationChannel.swift
//  bifrost
//
//  Created by Matheus Lutero on 23/02/21.
//

import Flutter
import Foundation

public typealias BifrostNotificationHandler = (_ arguments: Any?) -> Void

internal class BifrostNotificationChannel {
    
    private var notifications: Dictionary<String, BifrostNotificationHandler> = [:]
    private var channel: FlutterMethodChannel!
    
    init(_ messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(name: "bifrost/notification", binaryMessenger: messenger)
        channel.setMethodCallHandler(handle(_:result:))
    }
    
    func register(_ key: String, _ handler: @escaping BifrostNotificationHandler) {
        notifications[key] = handler
    }
    
    func unregister(_ key: String) {
        notifications.removeValue(forKey: key)
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let key = call.method
        let args = call.arguments
        
        if let notification = notifications[key] {
            notification(args); result(true)
        } else {
            result(false)
        }
    }
}
